### ISSUE TITLE
processor.php getHostUrl() does not detect the server port correctly

### ISSUE THREAD

**Issue by boldhedgehog** [magento/magento2#12969]
When Magento runs using the sandwich hosting set up (nginx-Varnish-nginx), and the backend nginx run on a not standard port (80 nor 443), the base URL is detected wrong in pub/errors/processor.hp

### Preconditions
1. Magento 2.1.7
2. nginx as the SSL offloader on 80, 443 ports
3. Varnish used for FPC
4. nginx as the backend for Varnish listening to not standard port (for example 8081)
5. PHP-FPM runs Magento
6. Production mode

### Steps to reproduce
1. Open any media resource that does not exist and is not allowed in `var/resource_config.json`

### Expected result
1. Magento's 404 page should be displayed in the browser with a link to configured skin CSS.
2. Base in the HTML code should be generated based on the requested host, for example: 
```html
<base href="https://domain.com/errors/default/" />
```

### Actual result
1. Magento's 404 page should be displayed in the browser with a link to configured skin CSS.
2. Base in the HTML code is generated based on the requested host *AND* `$_SERVER['SERVER_PORT']` , for example: 
```html
<base href="https://domain.com:8081/errors/default/" />
```
3. No static assets with relative path can be loaded due to the faulty base

The logic in `processor.php` should be similar to one in `//vendor/zendframework/zend-http/src/PhpEnvironment/Request.php`:
* Try to use HTTP_HOST header
* If the above is not available, the host is `$_SERVER['SERVER_NAME']`:`$_SERVER['SERVER_PORT']`

At this moment also in 2.3-develop branch the server port is always concatenated:

```php
public function getHostUrl()
    {
        /**
         * Define server http host
         */
        if (!empty($_SERVER['HTTP_HOST'])) {
            $host = $_SERVER['HTTP_HOST'];
        } elseif (!empty($_SERVER['SERVER_NAME'])) {
            $host = $_SERVER['SERVER_NAME'];
        } else {
            $host = 'localhost';
        }

        $isSecure = (!empty($_SERVER['HTTPS'])) && ($_SERVER['HTTPS'] != 'off');
        $url = ($isSecure ? 'https://' : 'http://') . $host;

// ------> should not be the case if the $host is assigned 
// ------> from  $_SERVER['HTTP_HOST']
       if (!empty($_SERVER['SERVER_PORT']) && !in_array($_SERVER['SERVER_PORT'], [80, 433])
            && !preg_match('/.*?\:[0-9]+$/', $url)
        ) {
            $url .= ':' . $_SERVER['SERVER_PORT'];
        }
        return  $url;
    }
```

**Comment by magento-engcom-team** [comment]
@boldhedgehog, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by sidolov** [comment]
Hi @boldhedgehog. Thank you for your report. 
The issue has been fixed in magento/magento2#18393 by @loganstellway in 2.3-develop branch
Related commit(s):
 - [422d24444321f7c5d5f6ae7fe7ec9872e233ba7a](https://github.com/magento/magento2/commit/422d24444321f7c5d5f6ae7fe7ec9872e233ba7a)
 - [1e28a8e25d1d43f02e090faf8aa77216a3a84738](https://github.com/magento/magento2/commit/1e28a8e25d1d43f02e090faf8aa77216a3a84738)
 - [7d921b60f87eecd591194777f2bd8ab59fd37dd5](https://github.com/magento/magento2/commit/7d921b60f87eecd591194777f2bd8ab59fd37dd5)

The fix will be available with the upcoming 2.3.1 release.

**Comment by magento-engcom-team** [comment]
Hi @boldhedgehog. Thank you for your report. 
The issue has been fixed in magento/magento2#18659 by @gelanivishal in 2.2-develop branch
Related commit(s):
 - [ef426a722fd4bb80470329f83d5a08f7a0424064](https://github.com/magento/magento2/commit/ef426a722fd4bb80470329f83d5a08f7a0424064)
 - [774a8a25688bd93cedc5b44cc79c9251a221d25b](https://github.com/magento/magento2/commit/774a8a25688bd93cedc5b44cc79c9251a221d25b)
 - [763be496a1c19e0008524a0eab8c1ae7554a2609](https://github.com/magento/magento2/commit/763be496a1c19e0008524a0eab8c1ae7554a2609)

The fix will be available with the upcoming 2.2.8 release.
